x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/BaseRunAsSuperuserCommand.java

```diff
@@ -0,0 +1,281 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.tool;
+
+import joptsimple.OptionSet;
+
+import org.elasticsearch.cli.ExitCodes;
+import org.elasticsearch.cli.KeyStoreAwareCommand;
+import org.elasticsearch.cli.Terminal;
+import org.elasticsearch.cli.UserException;
+import org.elasticsearch.common.io.Streams;
+import org.elasticsearch.common.settings.KeyStoreWrapper;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.CheckedFunction;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.xpack.core.XPackSettings;
+import org.elasticsearch.xpack.core.security.authc.support.Hasher;
+import org.elasticsearch.xpack.security.authc.file.FileUserPasswdStore;
+import org.elasticsearch.xpack.security.authc.file.FileUserRolesStore;
+import org.elasticsearch.xpack.security.support.FileAttributesChecker;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.Path;
+import java.security.SecureRandom;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * A {@link KeyStoreAwareCommand} that can be extended fpr any CLI tool that needs to allow a local user with
+ * filesystem write access to perform actions on the node as a superuser. It leverages temporary file realm users
+ * with a `superuser` role.
+ */
+public abstract class BaseRunAsSuperuserCommand extends KeyStoreAwareCommand {
+
+    private static final String[] ROLES = new String[] { "superuser" };
+    private static final int PASSWORD_LENGTH = 14;
+
+    private final Function<Environment, CommandLineHttpClient> clientFunction;
+    private final CheckedFunction<Environment, KeyStoreWrapper, Exception> keyStoreFunction;
+    private SecureString password;
+    private String username;
+    final SecureRandom secureRandom = new SecureRandom();
+
+    public BaseRunAsSuperuserCommand(
+        Function<Environment, CommandLineHttpClient> clientFunction,
+        CheckedFunction<Environment, KeyStoreWrapper, Exception> keyStoreFunction
+    ) {
+        super("description");
+        this.clientFunction = clientFunction;
+        this.keyStoreFunction = keyStoreFunction;
+    }
+
+    @Override
+    public void close() {
+        if (password != null) {
+            password.close();
+        }
+    }
+
+    @Override
+    protected final void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        validate(terminal, options, env);
+        KeyStoreWrapper keyStoreWrapper = keyStoreFunction.apply(env);
+        final Environment newEnv;
```

Comment:
Looks like you're building an environment here.
There is the `EnvironmentAwareCommand` to handle that.

Reply:
`BaseRunAsSuperuserCommand` extends `KeyStoreAwareCommand` which extends `EnvironmentAwareCommand` already. What I need to do here is to to factor in the secure settings we might have read from the keystore , thus the new environment. Not sure how `EnvironmentAwareCommand` can handle that for me ,let me know if I'm missing something !